libmachine/mcnutils/b2d.go

```diff
@@ -260,39 +259,122 @@ func (b *B2dUtils) CopyIsoToMachineDir(isoURL, machineName string) error {
+	if _, err := os.Stat(b.imgCachePath); os.IsNotExist(err) {
+		log.Infof("Image cache directory does not exist, creating it at %s...", b.imgCachePath)
+		if err := os.Mkdir(b.imgCachePath, 0700); err != nil {
 			return err
 		}
 	}
 
-	if err := CopyFile(b.commonIsoPath, machineIsoPath); err != nil {
+	apiURL := b.defaultURL
+	exists := pathExists(b.commonIsoPath)
+	latest := b.isLatestISO(apiURL)
+
+	if exists && latest {
+		log.Infof("Latest Boot2Docker ISO found locally, copying it to %s...", machineIsoPath)
+		return CopyFile(b.commonIsoPath, machineIsoPath)
+	}
+
+	if !exists {
+		log.Info("No default Boot2Docker ISO found locally, downloading the latest release...")
+	} else if !latest {
+		log.Info("Default Boot2Docker ISO is out-of-date, downloading the latest release...")
+	}
+	if err := b.DownloadLatestBoot2Docker(apiURL); err != nil {
 		return err
 	}
 
-	return nil
+	log.Infof("Copying %s to %s...", b.commonIsoPath, machineIsoPath)
+	return CopyFile(b.commonIsoPath, machineIsoPath)
+}
+
+// isoVer scans isopath and returns its version tag.
+func (b *B2dUtils) isoVer(isopath string) (string, error) {
+	iso, err := os.Open(isopath)
+	if err != nil {
+		return "", err
+	}
+	defer iso.Close()
+
+	isoMetadata := make([]byte, b.volumeIDLength)
+	_, err = iso.ReadAt(isoMetadata, b.volumeIDOffset)
+	if err != nil {
+		return "", err
+	}
+
+	verRegex := regexp.MustCompile(`v\d+\.\d+\.\d+`)
+	ver := string(verRegex.Find(isoMetadata))
+	log.Debug("local Boot2Docker ISO version: ", ver)
+	return ver, nil
+}
+
+// isLatestISO checks the latest release version tag from apiURL and
+// reports whether the local ISO cache is the latest version.
+//
+// It returns false if failing to get the local ISO version,
+// and true if failing to fetch the latest release version tag.
+func (b *B2dUtils) isLatestISO(apiURL string) bool {
+	localVer, err := b.isoVer(b.commonIsoPath)
+	if err != nil {
+		log.Warn("Unable to get the local Boot2Docker ISO version: ", err)
+		return false
+	}
+
+	latestVer, err := b.getReleaseTag(apiURL)
+	if err != nil {
+		log.Warn("Unable to get the latest Boot2Docker ISO release version: ", err)
+		return true
+	}
+
+	return localVer == latestVer
+}
+
+// getReleaseTag gets the latest release version tag from apiURL.
+// It assumes that the JSON response of apiURL has "tag_name" attribute.
+func (b *B2dUtils) getReleaseTag(apiURL string) (string, error) {
+	client := getClient()
+	req, err := b.getReleasesRequest(apiURL)
+	if err != nil {
+		return "", err
+	}
+	rsp, err := client.Do(req)
+	if err != nil {
+		return "", err
+	}
+	defer rsp.Body.Close()
+
+	var t struct {
+		TagName string `json:"tag_name"`
+	}
+	if err := json.NewDecoder(rsp.Body).Decode(&t); err != nil {
+		return "", err
+	}
+	if t.TagName == "" {
+		return "", errGitHubAPIResponse
+	}
+	return t.TagName, nil
+}
+
+// pathExists reports whether a file or directory specified by path exists.
+func pathExists(path string) bool {
```

Comment:
For instance, this could become a method on aj interface and embedded into the `B2DUtils` struct:

```go
type PathChecker interface {
    PathExists(string) bool
}

type StatPathChecker struct {}

func (spc *StatPathChecker) PathExists(path string) bool {
    _, err := os.Stat(path)
    return os.IsNotExist(err)
}

type B2DUtils struct {
    StatPathChecker
    ...
}
```